### PR TITLE
adds crdNoOverlappingGVRAdmission admission 

### PR DESCRIPTION
**Pull request by p0lyn0mial**
## Summary
introduces an admission plugin for ensuring CRD's Group and Resource fields for logical clusters are unique and don't overlap with bound CRDs.

TODO:
- change the binding controller to check for conflicting GRVs (separate PR)
## Related issue(s)

xref: https://github.com/kcp-dev/kcp/issues/748

### CODE REVIEW

File: pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+		if err != nil {
+			return err
+		}
+
+		for _, apiExportSchemaName := range apiExport.Spec.LatestResourceSchemas {
+			apiResourceSchema, err := p.getAPIResourceSchema(apiExportClusterName, apiExportSchemaName)
+			if err != nil {
+				return err
+			}
+			if apiResourceSchema.Spec.Group == crd.Spec.Group && apiResourceSchema.Spec.Names.Plural == crd.Spec.Names.Plural {
```
Comment (p0lyn0mial) [inline pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:93]: I think comparing `Group` and `Resource` is enough, we don't have to take `Versions` into consideration.

File: pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	if err != nil {
+		return err
+	}
+	for _, apiBindingForCurrentClusterName := range apiBindingsForCurrentClusterName {
+		apiExportClusterName, err := getAPIExportClusterName(apiBindingForCurrentClusterName)
+		if err != nil {
+			// since all exports needs to be checked we need to stop in case of an error
+			return err
+		}
+		apiExport, err := p.getAPIExport(apiExportClusterName, apiBindingForCurrentClusterName.Spec.Reference.Workspace.ExportName)
```
Comment (p0lyn0mial) [inline pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:83]: I think that `CRDs` from an APIExport are the ones that are eventually created in the "current" cluster.

File: pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+			informers.Apis().V1alpha1().APIResourceSchemas().Informer().HasSynced()
+	})
+	p.apiBindingLister = informers.Apis().V1alpha1().APIBindings().Lister()
+	p.apiExportLister = informers.Apis().V1alpha1().APIExports().Lister()
+	p.apiResourceSchemaLister = informers.Apis().V1alpha1().APIResourceSchemas().Lister()
+}
+
+// Validate checks if the given CRD's Group and Resource don't overlap with other CRDs exposed via APIBindings
+func (p *crdNoOverlappingGVRAdmission) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	if a.GetResource().GroupResource() != apiextensions.Resource("customresourcedefinitions") {
```
Comment (p0lyn0mial) [inline pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:57]: I think I need to find a way to skip `CRDs` created by the APIBinding controller

Comment (p0lyn0mial) [reply pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:57]: or actually, we should make a distinction between CRDs created for the binding and the ones created by an admin

File: pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go (hunk 1 above)
Comment (sttts) [reply pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:93]: yes, all versions should be owned either by the CRD or the binding.

File: pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go (hunk 2 above)
Comment (sttts) [reply pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:83]: try to not depend on the APIExport. The APIBinding itself (the status) must be enough for this decision.

Background: APIExport might live on a different shard and we should not break if the shard is unavailable (network partition). If something is missing for the logic in the APIBinding status, we should add it.

File: pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go (hunk 3 above)
Comment (sttts) [reply pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:57]: the ones from APIBinding live in the `system:bounds-crds` logical cluster.

File: pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:83]: oh, ic, does it mean we should reject the request if APIBinding is still progressing?

Comment (sttts) [reply pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:83]: The status of the APIBinding is what counts. This admission plugin also has to stop binding to succeed (with some good condition message).

File: pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go (hunk 3 above)
Comment (p0lyn0mial) [reply pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:57]: there seem to be no way of identifying the target cluster from a CRD in the logical cluster

Comment (sttts) [reply pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:57]: I think the APIBinding CRDs in `system:bound-crds` you can ignore here as long as you have the GVRs for them from APIBinding.status.

The `system:admin` CRDs matter too here (today that's all from https://github.com/kcp-dev/kcp/tree/main/config/crds, but most of them will turn into APIBindings soon).

File: pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	}
+	if a.GetKind().GroupKind() != apiextensions.Kind("CustomResourceDefinition") {
+		return nil
+	}
+	clusterName, err := request.ClusterNameFrom(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster from context: %w", err)
+	}
+	// ignore CRDs targeting system workspaces
+	if clusterName == apibinding.ShadowWorkspaceName || clusterName == logicalcluster.New("system:admin") {
```
Comment (sttts) [inline pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:61]: maybe even `!clusterName.HasPrefix(tenancyv1alpha1.RootCluster)`

File: pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+		}
+	}
+	if len(notReadyAPIBindings) > 0 {
+		return admission.NewForbidden(a, fmt.Errorf("cannot create %q crd because the following apiBindings werern't ready and werern't checked for conflicting GVRs = %v", crd.Name, notReadyAPIBindings))
+	}
+	return nil
+}
+
+func (p *crdNoOverlappingGVRAdmission) listAPIBindingsFor(clusterName logicalcluster.Name) ([]*apisv1alpha1.APIBinding, error) {
+	list, err := p.apiBindingLister.List(labels.Everything())
```
Comment (sttts) [inline pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:93]: use an index as in https://github.com/kcp-dev/kcp/blob/main/pkg/reconciler/scheduling/placement/placement_indexes.go#L30.

File: pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+		}
+		for _, boundResource := range apiBindingForCurrentClusterName.Status.BoundResources {
+			if boundResource.Group == crd.Spec.Group && boundResource.Resource == crd.Spec.Names.Plural {
+				return admission.NewForbidden(a, fmt.Errorf("cannot create %q crd with %q group and %q resource because it overlaps with a bound crd for %q apiBinding in %q logical cluster",
+					crd.Name, crd.Spec.Group, crd.Spec.Names.Plural, apiBindingForCurrentClusterName.Name, clusterName))
+			}
+		}
+	}
+	if len(notReadyAPIBindings) > 0 {
+		return admission.NewForbidden(a, fmt.Errorf("cannot create %q crd because the following apiBindings werern't ready and werern't checked for conflicting GVRs = %v", crd.Name, notReadyAPIBindings))
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:87]: why fail at all in this case?

File: pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+	}
+	if a.GetKind().GroupKind() != apiextensions.Kind("CustomResourceDefinition") {
+		return nil
+	}
+	clusterName, err := request.ClusterNameFrom(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster from context: %w", err)
+	}
+	// ignore CRDs targeting system and non-root workspaces
+	if clusterName == apibinding.ShadowWorkspaceName || clusterName == logicalcluster.New("system:admin") || !clusterName.HasPrefix(tenancyv1alpha1.RootCluster) {
```
Comment (sttts) [inline pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:99]: nit: the last test implies the other two. Am fine with either, the first two or the last :)

File: pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+		for _, boundResource := range apiBindingForCurrentClusterName.Status.BoundResources {
+			if boundResource.Group == crd.Spec.Group && boundResource.Resource == crd.Spec.Names.Plural {
+				return admission.NewForbidden(a, fmt.Errorf("cannot create %q CustomResourceDefinition with %q group and %q resource because it overlaps with a bound CustomResourceDefinition for %q APIBinding in %q logical cluster",
+					crd.Name, crd.Spec.Group, crd.Spec.Names.Plural, apiBindingForCurrentClusterName.Name, clusterName))
+			}
+		}
+	}
+	if len(notReadyAPIBindings) > 0 {
+		return admission.NewForbidden(a, fmt.Errorf("cannot create %q CustomResourceDefinition because the following APIBinding werern't ready and werern't checked for conflicting GVRs = %v", crd.Name, notReadyAPIBindings))
+	}
```
Comment (sttts) [inline pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:126]: still think this whole if clause is not right. Being in !BindingUpToDate is normal, not a reason to stop CRD creation.

Comment (p0lyn0mial) [reply pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:126]: seems like we need to trade off availability over consistency. My intention was to make sure we have seen all bounded CRDs before admitting a CRD. I guess you prefer the former or is there any other condition we should check? (`!InitialBindingCompleted`?)

Comment (sttts) [reply pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:126]: resources not bound via APIBinding status do not exist :)

Comment (sttts) [reply pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go:126]: Same with `!InitialBindingCompleted`. At that point there are no bound resources yet. So no conflict.

There is the second part with the APIBinding controller that it needs similar logic for the inverse check.

Comment (p0lyn0mial) [reply pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go]: yeah, at this point it is a best-effort admission

**Comment by p0lyn0mial** [comment]
@sttts updated, PTAL

**Comment by p0lyn0mial** [comment]
run it locally and it gave me `error: plugins [apis.kcp.dev/CRDNoOverlappingGVR] registered are not in RecommendedPluginOrder`

**Pull request merged**